wgsl/index.bs

```diff
@@ -395,44 +400,145 @@ The following types are <dfn dfn noexport>storable</dfn>:
+* [[#array-types]] if its element type is IO-shareable, and the array is not [=runtime-sized=]
+* [[#struct-types]] if all its members are IO-shareable
+
+The following kinds of values must be of IO-shareable type:
+
+* Values read from or written to built-in variables.
+* Values accepted as inputs from an upstream pipeline stage.
+* Values written as output for downstream processing in the pipeline, or to an output attachment.
+
 ### Host-shareable Types ### {#host-shareable-types}
 
 The following types are <dfn dfn noexport>host-shareable</dfn>:
 
 * [=numeric scalar=] types
 * [=numeric vector=] types
 * [[#matrix-types]]
-* [[#array-types]] if the array type has a stride attribute and its element type is host-shareable
-* [[#struct-types]] if all its members are host-shareable
+* [[#array-types]] if the array type has a [=stride=] attribute and its element type is host-shareable
+* [[#struct-types]] if each of its members both has an [=offset=] attribute and is host-shareable
+
+Host-shareable types are used to describe the contents of buffers which are shared between
+the host and the GPU, or copied between host and GPU without format translation.
+When used for this purpose, the type must be additionally decorated with layout attributes
+as described in [[#memory-layout]].
+In particular, any variable using the `uniform` or `buffer` [=storage classes=] must have
+a host-shareable [=store type=].
+
+Note: An [=IO-shareable=] type would also be host-shareable if it and its subtypes have
+the approporate [=stride=] and [=offset=] attributes.
+Additionally, a [=runtime-sized=] array is host-shareable but is not IO-shareable.
+
+Note: Both IO-shareable and host-shareable types have concrete sizes, but counted differently.
+IO-shareable types are sized by a location-count metric, see [[#input-output-locations]].
+Host-shareable types are sized by a byte-count metric, see [[#memory-layout]].
 
-TODO: *This is a stub*: Collectively, the *stride* and *offset* attributes are called *layout attributes*.
+### Storage Classes ### {#storage-class}
 
-### Storage Classes TODO ### {#storage-class}
+Memory locations are partitioned into <dfn noexport>storage classes</dfn>.
+Each storage class has unique properties determining
+mutability, visibility, the values it may contain,
+and how to use variables with it.
+
+<table class='data'>
+  <thead>
+    <tr><td>Storage class
+        <td>Readable by shader?<br>Writable by shader?
+        <td>Sharing among invocations
+        <td>Variable scope
+        <td>Restrictions on stored values
+        <td>Notes
+  </thead>
+  <tr><td>in
+      <td>Read-only
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Input from an upstream pipeline stage, or from the implementation.
+  <tr><td>out
+      <td>Read-write
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Output to a downstream pipeline stage.
+  <tr><td>function
```

Comment:
Is this useful? Do we have a situation where this storage class isn't known apriori?

Reply:
We need a name for the concept. This table introduces the storage classes, so function storage class should exist hee.

Also, we'll be able to form pointer-to-function values.  (Note: that part isn't yet written in the spec.)